### PR TITLE
i argument to vec_slice() gains NULL default

### LINKED ISSUE #246: Do we need/have an equivalent for x[] <- value?
Required when we start forwarding `[<-()` to `vec_slice()` .

``` r
library(vctrs)

x <- 1:3
x[] <- 4L

x <- 1:3
vec_slice(x, ) <- 4L
#> Error in vec_size(i): argument "i" is missing, with no default
```

<sup>Created on 2019-03-24 by the [reprex package](https://reprex.tidyverse.org) (v0.2.1.9000)</sup>

### PR DESCRIPTION
**Pull request by krlmlr**
I'm sure this can also be fixed using missing arguments, but explicit `NULL`s seem to be much easier to reason about.

Conflicts with #249, but easy to redo on top of that.

Closes #246.

### CODE REVIEW

File: src/slice.c [hunk 1]
```diff
@@ -352,11 +352,11 @@ static SEXP chr_as_index(SEXP i, SEXP x) {
 
 SEXP vec_as_index(SEXP i, SEXP x) {
   switch (TYPEOF(i)) {
+  case NILSXP: return lgl_as_index(Rf_ScalarLogical(1), x);
```
Comment (lionel-) [inline src/slice.c]: This needs protection. Perhaps in `nil_as_index()` so the protection/unprotection doesn't mess up the switch.

Comment (lionel-) [reply src/slice.c]: (I don't want to rely on the fact thet scalarlogical returns a shared constant)

Oh this can just be `vctrs_shared_true`.

**Comment by lionel-** [comment]
@krlmlr Can you make `i` default to `TRUE` please?

**Comment by krlmlr** [comment]
Do we also want to replicate R's behavior for `is.null(i)` ?

**Comment by lionel-** [comment]
> Do we also want to replicate R's behavior for is.null(i) ?

Sorry I don't understand the question.

**Comment by krlmlr** [comment]
Do we need this for `vec_slice()`:

``` r
iris[NULL, ]
#> [1] Sepal.Length Sepal.Width  Petal.Length Petal.Width  Species     
#> <0 rows> (or 0-length row.names)
letters[NULL]
#> character(0)
letters[NULL] <- "A"
letters
#>  [1] "a" "b" "c" "d" "e" "f" "g" "h" "i" "j" "k" "l" "m" "n" "o" "p" "q"
#> [18] "r" "s" "t" "u" "v" "w" "x" "y" "z"
```

<sup>Created on 2019-03-26 by the [reprex package](https://reprex.tidyverse.org) (v0.2.1.9000)</sup>

**Comment by hadley** [comment]
No

**Comment by krlmlr** [comment]
Then I'd rather leave the default as `NULL`, that's the desired semantics (#115).

**Comment by lionel-** [comment]
Your question is more general. Do we need this for vctrs:

```r
letters[integer()] <- "A"
letters
#>  [1] "a" "b" "c" "d" "e" "f" "g" "h" "i" "j" "k" "l" "m" "n" "o" "p" "q" "r"
#> [19] "s" "t" "u" "v" "w" "x" "y" "z"
```

Again this has compositional implications that should be carefully thought out.

I strongly recommend against `foo[c(idx1, idx2)]` having unstable semantics.

**Comment by lionel-** [comment]
Maybe it'd make sense to have `NULL` as default with the current semantics. That would be the most consistent with #115 because `foo[bar]` would keep obs in `bar` and `foo[]` wouldn't keep any obs. But that'd make `[` methods powered by `vec_slice()` different from base R's. You'd need to use `TRUE` instead of a missing index.

I think base R's behaviour for missing indices makes sense in the context of multi-dim subsetting, where `[, ]` means subset all rows and all columns. So we can't really change that.

**Comment by krlmlr** [comment]
I agree that the following invariant should hold also for `NULL` values in `i1` and `i2`:

```r
identical(vec_c(x[i1], x[i2]), x[vec_c(i1, i2)])
```

But it feels even more important that `x[]` should be the same as `vec_slice(x, )` . This means that we need to deviate from the "`NULL` means missing" idiom in this particular case. I think it's fine if we document it in #115.

**Comment by lionel-** [comment]
Here is what I suggest:

- Avoid missing arg in `vec_slice()` and `vec_slice<-` because they're for programmatic use.

- Keep missing arg in `[`. Interface with `vec_slice()` with `maybe_missing(i, default = )`.

- For `[`, the default missing arg is translated to `TRUE`. For `[<-`, it is translated to `NULL`.

- The defaults for `vec_slice()` and `vec_slice<-` are both `NULL`, which is equivalent to an empty index.

This way `vec_slice(x)` and `vec_slice(x, int())`, as well as `vec_slice(x) <- foo` and `vec_slice(x, int()) <- foo`, have equivalent semantics. Not supplying an index and supplying an empty index are the same. As for our `[` methods, their behaviour is consistent in base R and with multi-dimensional subsetting.

**Comment by lionel-** [comment]
> it feels even more important that x[] should be the same as vec_slice(x, )

It doesn't seem important to me.

**Comment by krlmlr** [comment]
What do we gain by making `vec_slice(i, )` behave differently to `x[]` ?

**Comment by lionel-** [comment]
Consistency. An absence of index and an empty index are the same.

`vec_slice()` is for programmatic use and one-dimensional subsetting. It doesn't have the same interface constraints as `[`.

In fact maybe a missing index should just be an error for `vec_slice()` and `vec_slice<-`. The caller should be forced to use the appropriate sentinel (e.g. `TRUE` for all obs, and `NULL` / `int()` for no obs).

**Comment by krlmlr** [comment]
Fully agree, implemented this way.

**Comment by hadley** [comment]
Should we make `vec_slice(x, NULL)` our standard approach to generate a size-0 slice?

**Comment by lionel-** [comment]
> Should we make vec_slice(x, NULL) our standard approach to generate a size-0 slice?

Perhaps `vec_slice(x, int())` is a bit easier to understand for people not familiar with the syntax.

**Pull request merged**